### PR TITLE
[Blazor] Fix type name hashing when the type has multibyte characters

### LINKED ISSUE #50879: ComputeComponentTypeNameHash does not handle multi-byte type names correctly
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

👋 I was looking though the use of cryptography in ASP.NET Core, and stumbled on this code block:

https://github.com/dotnet/aspnetcore/blob/46562b1435bf111a7425b40f507b157b42a016a4/src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs#L206-L213

Nothing about the use of cryptography there concerns me too much, but I did want to point out that it does not handle types where the type name requires the use of multibyte characters.

I lack detailed knowledge of how SSR works, so I had a bit of trouble trying to reproduce it in a real app.

### Expected Behavior

_No response_

### Steps To Reproduce

I had a bit of trouble trying to reproduce this using an app, as I am not super familiar with the innards of SSR. However, in isolation it can be reproduced with:

```C#
using System;
using System.Text;
using System.Security.Cryptography;

string hash = ComputeComponentTypeNameHash(typeof(こんにちは));
Console.WriteLine(hash);

static string ComputeComponentTypeNameHash(Type componentType)
{
    if (componentType.FullName is not { } typeName)
    {
        throw new InvalidOperationException($"An invalid component type was used.");
    }

    var typeNameLength = typeName.Length;
    var typeNameBytes = typeNameLength < 1024
        ? stackalloc byte[typeNameLength]
        : new byte[typeNameLength];

    Encoding.UTF8.GetBytes(typeName, typeNameBytes);

    Span<byte> typeNameHashBytes = stackalloc byte[SHA1.HashSizeInBytes];
    SHA1.HashData(typeNameBytes, typeNameHashBytes);

    return Convert.ToHexString(typeNameHashBytes);
}

class こんにちは
{

}
```

### Exceptions (if any)

This will produce a `System.ArgumentException` since the destination buffer is too small.

### .NET Version

_No response_

### Anything else?

Please feel free to close this issue if the type in `ComputeComponentTypeNameHash` is not exposed to developers.

### LINKED ISSUE #52109: Blazor Web App Run Error when project name contain Chinese
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

if the project name contain any Chinese, the counter Page will raise error.
Snipaste_2023-11-16_08-44-27

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Exceptions (if any)

_No response_

### .NET Version

8.0.100

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by MackinnonBuck**
# Fix type name hashing when the type has multibyte characters

Fixes an issue where an exception gets thrown if a render mode boundary component has a type whose full name contains multibyte characters.

## Description

This PR fixes the issue by allocating a constant-sized stack buffer and falling back to a heap-allocated buffer when the type name is too long.

Fixes #50879
Fixes #52109

### CODE REVIEW

**Comment by MackinnonBuck** [comment]
I'll create a backport to `release/8.0` for servicing after this gets approved/merged

File: src/Components/Endpoints/test/TypeNameHashTest.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+public class TypeNameHashTest
+{
+    // In these tests, we're mostly interested in checking that the hash function succeeds
+    // for any type with a valid name. We'll also do some basic sanity checking by ensuring
+    // that the string representation of the hash has the expected length.
+
+    // We currently use a hex-encoded SHA1 hash, so there should be two characters per byte
+    // of encoded data.
+    private const int ExpectedHashLength = SHA1.HashSizeInBytes * 2;
```
Comment (surayya-MS) [inline src/Components/Endpoints/test/TypeNameHashTest.cs:16]: @javiercn has mentioned in one of our conversations that we should also update from SHA1 to SHA256

**Review by surayya-MS — APPROVED** [review]
Looks great!

File: src/Components/Endpoints/test/TypeNameHashTest.cs (hunk 1 above)
Comment (javiercn) [reply src/Components/Endpoints/test/TypeNameHashTest.cs:16]: Yes, please.

**Review by javiercn — APPROVED** [review]

**Pull request merged**